perfkitbenchmarker/providers/docker/docker_virtual_machine.py

```diff
@@ -0,0 +1,303 @@
+# Copyright 2018 PerfKitBenchmarker Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Contains code related to lifecycle management of Docker Containers."""
+
+import json
+import logging
+import os
+import threading
+
+from perfkitbenchmarker import data
+from perfkitbenchmarker import context
+from perfkitbenchmarker import disk
+from perfkitbenchmarker import errors
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import providers
+from perfkitbenchmarker import virtual_machine, linux_virtual_machine
+from perfkitbenchmarker import vm_util
+from perfkitbenchmarker.providers.docker import docker_disk
+from perfkitbenchmarker.providers.docker import docker_resource_spec
+from perfkitbenchmarker.vm_util import OUTPUT_STDOUT as STDOUT
+
+FLAGS = flags.FLAGS
+
+UBUNTU_IMAGE = 'ubuntu:xenial'
+DEFAULT_DOCKER_IMAGE = 'pkb/ubuntu16_ssh'
+
+
+class DockerVirtualMachine(virtual_machine.BaseVirtualMachine):
+  """
+  Object representing a Docker Container instance
+  """
+  CLOUD = providers.DOCKER
+  DEFAULT_IMAGE = None
+  CONTAINER_COMMAND = None
+  docker_build_lock = threading.Lock()
+
+  def __init__(self, vm_spec):
+    """Initialize a Docker Container.
+
+    Args:
+      vm_spec
+    """
+    super(DockerVirtualMachine, self).__init__(vm_spec)
+    self.name = self.name.replace('_', '-')
+    self.container_id = ''
+    self.user_name = FLAGS.username
+    self.image = self.image or self.DEFAULT_IMAGE
+    self.cpus = vm_spec.docker_provider_cpus
+    self.memory_mb = vm_spec.docker_provider_memory_mb
+    self.privileged = vm_spec.privileged_docker
+    self.container_image = DEFAULT_DOCKER_IMAGE
+    self.docker_sysctl_flags = ''
+    # apply flags
+    if FLAGS.docker_custom_image:
+      self.container_image = FLAGS.docker_custom_image
+    if FLAGS.docker_sysctl_flags:
+      self.docker_sysctl_flags = FLAGS.docker_sysctl_flags
+
+  def _CreateDependencies(self):
+    self._CreateVolumes()
+
+  def _DeleteDependencies(self):
+    self._DeleteVolumes()
+
+  def _Create(self):
+    """Create a Docker instance"""
+
+    # Locally build docker container
+    with self.docker_build_lock:
+      image_exists = self._LocalImageExists(self.container_image)
+      if image_exists is False:
+        self._BuildImageLocally()
```

Comment:
Is the docker image read-only? I'm curious about the case where we might have two Docker virtual machines. Is there any reason to support that?

Reply:
A docker image is read-only. This bit of code is to prevent a specific case where Docker image doesn't yet exist on the local machine and you are running a benchmark that uses two containers, then it will attempt to build the Docker image in two threads at the same time, leading to an untagged duplicate image that is never used and just taking up space.

My specific use case for using multiple docker containers was when I was developing/testing a networking benchmark and containers were faster to test on. Other than testing, inter-container communication could be a reasonable thing to benchmark